ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1527 @@
+/* Byte offset into multi-byte encoded string -> number of complete characters up to that position
+ * Negative offsets count back from the end of the string, but cannot count back further than the start of the string
+ * Positive offsets which are greater than the length of the string are clamped to the end of the string */
+static long byte_offset_to_char_offset(const mbfl_encoding *encoding, unsigned char *start, size_t length, ssize_t byte_offset)
+{
+	if (byte_offset == 0) {
+		return 0;
+	} else if (byte_offset < 0) {
+		byte_offset += length;
+		if (byte_offset < 0) {
+			/* negative offsets greater than length of string are clamped to zero */
+			return 0;
+		}
+	} else if (byte_offset > length) {
+		byte_offset = length;
+	}
+
+	/* If encoding uses a constant number of bytes per character, we can trivially convert
+	 * byte offset to character offset */
+	if (encoding->flag & MBFL_ENCTYPE_SBCS) {
+		return byte_offset;
+	} else if (encoding->flag & (MBFL_ENCTYPE_WCS2BE | MBFL_ENCTYPE_WCS2LE)) {
+		return byte_offset / 2;
+	} else if (encoding->flag & (MBFL_ENCTYPE_WCS4BE | MBFL_ENCTYPE_WCS4LE)) {
+		return byte_offset / 4;
+	} else if (encoding->mblen_table != NULL) {
+		/* If not, can we determine the number of bytes for each character just by
+		 * looking at the first byte? If so, we can run through the string, using
+		 * a table lookup on the first byte to see how many bytes to skip ahead */
+		const unsigned char *mbtab = encoding->mblen_table;
+		unsigned char *p = start, *e = p + byte_offset;
+		size_t len = 0;
+		while (p < e) {
+			p += mbtab[*p];
+			len++;
+		}
+		if (p > e) {
+			len--;
+		}
+		return len;
+	} else {
+		/* Otherwise, do it the slow way: convert the string to wchar, and count those  */
+		size_t len = 0;
+		mbfl_convert_filter *filter = mbfl_convert_filter_new(encoding, &mbfl_encoding_wchar, filter_count_output, 0, &len);
+		mbfl_convert_filter_feed_string(filter, start, byte_offset);
+		mbfl_convert_filter_delete(filter);
+		return len;
+	}
+}
+
+/* Character offset into multi-byte encoded string -> number of bytes up to that position
+ * If requested offset is greater than character length of string, returns byte length of string
+ * Negative offsets count back from the end of the string, but cannot count past the beginning */
+static size_t char_offset_to_byte_offset(const mbfl_encoding *encoding, unsigned char *start, size_t length, ssize_t char_offset)
+{
+	if (char_offset == 0) {
+		return 0;
+	} else if (char_offset < 0) {
+		/* Need to go backwards from end of string */
+		char_offset = -char_offset;
+		if (char_offset >= length) {
+			/* Each character takes _at least_ one byte, so this will definitely go
+			 * to the beginning of the string */
+			return 0;
+		}
+		if (encoding->flag & MBFL_ENCTYPE_SBCS) {
+			return MAX((long)length - char_offset, 0);
+		} else if (encoding->flag & (MBFL_ENCTYPE_WCS2BE | MBFL_ENCTYPE_WCS2LE)) {
+			return MAX((long)length - (char_offset << 1), 0);
```

Comment:
```suggestion
			return MAX((long)length - (char_offset * 2), 0);
```
Using a different style here than for the same cases below :)